### PR TITLE
#490 - Assign externally managed Arakoon clusters to dedicated Backends

### PR DESCRIPTION
**Pull request by kvanhijf**

### CODE REVIEW

File: ovs/lib/alba.py [hunk 1]
```diff
@@ -1114,260 +1115,287 @@ def _validate_asd_removal(storage_router_ip):
...
+        if len(nsm_cluster_names) > 0:
             if alba_backend_guid is None:
                 raise ValueError('Additional NSMs can only be configured for a specific ALBA Backend')
-            if not isinstance(additional_nsms, dict):
-                raise ValueError("'additional_nsms' must be of type 'dict'")
-            if not isinstance(additional_nsm_names, list):
-                raise ValueError("'additional_nsm_names' must be of type 'list'")
-            if additional_nsm_amount is None or not isinstance(additional_nsm_amount, int):
-                raise ValueError('Amount of additional NSM clusters to deploy must be specified and 0 or more')
+            if not isinstance(nsm_cluster_names, list):
```
Comment (JeffreyDevloo) [inline ovs/lib/alba.py]: Duplicate check?

File: ovs/lib/alba.py [hunk 2]
```diff
@@ -1114,260 +1115,287 @@ def _validate_asd_removal(storage_router_ip):
...
 
-        :param allow_offline: Ignore offline nodes
-        :type allow_offline: bool
         :param alba_backend_guid: Run for a specific ALBA Backend
         :type alba_backend_guid: str
         :param min_nsms: Minimum amount of NSM hosts that need to be provided
         :type min_nsms: int
-        :param additional_nsms: Information about the additional clusters to claim (and create for internally managed Arakoon clusters)
-        :type additional_nsms: dict
+        :param nsm_cluster_names: Information about the additional clusters to claim (only for externally managed Arakoon clusters)
```
Comment (JeffreyDevloo) [inline ovs/lib/alba.py]: Indicate that is it for an externally managed nsm in its name

File: ovs/lib/alba.py [hunk 3]
```diff
@@ -1114,260 +1115,287 @@ def _validate_asd_removal(storage_router_ip):
...
+                        nsms_to_add = len(nsm_cluster_names)  # For externally managed clusters we only claim the specified clusters, if none provided, we just log it
+                        if nsms_to_add == 0:
+                            AlbaController._logger.critical('ALBA Backend {0} - All NSM clusters are overloaded'.format(alba_backend.name))
+                            continue
+
+                # Deploy new (internal) or claim existing (external) NSM clusters
+                AlbaController._logger.debug('ALBA Backend {0} - Currently {1} NSM cluster{2}'.format(alba_backend.name, len(nsm_loads), '' if len(nsm_loads) == 1 else 's'))
+                AlbaController._logger.debug('ALBA Backend {0} - Trying to add {1} NSM cluster{2}'.format(alba_backend.name, nsms_to_add, '' if nsms_to_add == 1 else 's'))
                 base_number = max(nsm_loads.keys()) + 1
                 for count, number in enumerate(xrange(base_number, base_number + nsms_to_add)):
```
Comment (JeffreyDevloo) [inline ovs/lib/alba.py]: Count is an index here

File: ovs/lib/alba.py [hunk 4]
```diff
@@ -1114,260 +1115,287 @@ def _validate_asd_removal(storage_router_ip):
...
+                    else:
+                        nsms_to_add = len(nsm_cluster_names)  # For externally managed clusters we only claim the specified clusters, if none provided, we just log it
+                    if nsms_to_add == 0:
+                        continue
+                else:
+                    AlbaController._logger.warning('ALBA Backend {0} - NSM load is NOT OK'.format(alba_backend.name))
+                    if internal is True:
+                        nsms_to_add = max(1, min_nsms - len(nsm_loads))  # When load is not OK, deploy at least 1 additional NSM
+                    else:
+                        nsms_to_add = len(nsm_cluster_names)  # For externally managed clusters we only claim the specified clusters, if none provided, we just log it
```
Comment (JeffreyDevloo) [inline ovs/lib/alba.py]: Shouldn't the code check if the supplied external nsms are sufficient to add? Might be better to lo gsuch a thing
So if nsms_to_add  < len(nsm_cluster_names): log something about not having enough external nsms supplied to satisfy the current safety

Comment (kvanhijf) [reply ovs/lib/alba.py]: These are the clusters that will be claimed by the checkup. At this point they have been verified to be available AND not in use

**Pull request merged**